### PR TITLE
fix the num_pos bug in fcos

### PR DESCRIPTION
**Pull request by hyz-xmaster**
synchronize num_pos in fcos_head and change requires_grad=True for fcos

### CODE REVIEW

**Comment by hyz-xmaster** [comment]
This PR is linked to this [issue](https://github.com/open-mmlab/mmdetection/issues/2933)

File: configs/fcos/fcos_r50_caffe_fpn_4x4_1x_coco.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
         num_stages=4,
         out_indices=(0, 1, 2, 3),
         frozen_stages=1,
-        norm_cfg=dict(type='BN', requires_grad=False),
+        norm_cfg=dict(type='BN', requires_grad=True),
```
Comment (yhcao6) [inline configs/fcos/fcos_r50_caffe_fpn_4x4_1x_coco.py:15]: Why do you modify this?

Comment (yhcao6) [reply configs/fcos/fcos_r50_caffe_fpn_4x4_1x_coco.py:15]: The `samples_per_gpu` and `workers_per_gpu` should change to 2 now.

**Comment by yhcao6** [comment]
Thanks for your PR to fix the issue.

File: configs/fcos/fcos_r50_caffe_fpn_4x4_1x_coco.py (hunk 1 above)
Comment (hyz-xmaster) [reply configs/fcos/fcos_r50_caffe_fpn_4x4_1x_coco.py:15]: Recently, I have read the mmdetection paper and noticed that the requires_grad=True is the default setting in mmdetection. I thought that this requirese_grad being set as False for FCOS only was due to the async num_pos issue. So I changed requires_grad=True after fixing the num_pos bug.

**But**, after running an initial experiment, I found that setting requires_grad=True hurt the performance for over 1 mAP.    I am running another experiment with requires=False to see the performance. If the performance is good, that means setting requires_grad=True is not a good choice and I will change it back to False. Otherwise, I need to debug the code further.

Comment (hellock) [reply configs/fcos/fcos_r50_caffe_fpn_4x4_1x_coco.py:15]: It depends on the pretrained backbone you use. In the paper we use pytorch style backbones and pretrained weights from torchvision. However, `open-mmlab://detectron/resnet50_caffe` is not a normal pretrained weight, which fused conv and bn layers and the weight/bias of bn layers are 0 and 1. Thus we should not set `requires_grad=True`.

Comment (hyz-xmaster) [reply configs/fcos/fcos_r50_caffe_fpn_4x4_1x_coco.py:15]: That makes sense and my experiment results also agree with this. I will change it back and do some minor changes to the over-length lines.

**Comment by hyz-xmaster** [comment]
Any idea of fixing this format[ issue](https://travis-ci.com/github/open-mmlab/mmdetection/jobs/346428268):

--- mmdet/models/dense_heads/fcos_head.py	(original)
+++ mmdet/models/dense_heads/fcos_head.py	(reformatted)
@@ -196,9 +196,7 @@
 ```
num_pos_avg_per_gpu = max(num_pos_avg_per_gpu, 1.0)
 
         loss_cls = self.loss_cls(
-            flatten_cls_scores,
-            flatten_labels,
-            avg_factor=num_pos_avg_per_gpu)
+            flatten_cls_scores, flatten_labels, avg_factor=num_pos_avg_per_gpu)
 
         pos_bbox_preds = flatten_bbox_preds[pos_inds]
         pos_centerness = flatten_centerness[pos_inds]
```
The command "yapf -r -d --style .style.yapf mmdet/ tools/ tests/ configs/" failed and exited with 1 during .

**Comment by yhcao6** [comment]
> Any idea of fixing this format[ issue](https://travis-ci.com/github/open-mmlab/mmdetection/jobs/346428268):
> 
> --- mmdet/models/dense_heads/fcos_head.py (original)
> +++ mmdet/models/dense_heads/fcos_head.py (reformatted)
> @@ -196,9 +196,7 @@
> 
> ```
> num_pos_avg_per_gpu = max(num_pos_avg_per_gpu, 1.0)
> 
>         loss_cls = self.loss_cls(
> -            flatten_cls_scores,
> -            flatten_labels,
> -            avg_factor=num_pos_avg_per_gpu)
> +            flatten_cls_scores, flatten_labels, avg_factor=num_pos_avg_per_gpu)
> 
>         pos_bbox_preds = flatten_bbox_preds[pos_inds]
>         pos_centerness = flatten_centerness[pos_inds]
> ```
> 
> The command "yapf -r -d --style .style.yapf mmdet/ tools/ tests/ configs/" failed and exited with 1 during .

The error tells you how to reformat the code.
You should change
```
loss_cls = self.loss_cls(
             flatten_cls_scores,
             flatten_labels,
             avg_factor=num_pos_avg_per_gpu)
```
to 
```
loss_cls = self.loss_cls(
       flatten_cls_scores, flatten_labels, avg_factor=num_pos_avg_per_gpu)
```

**Comment by hyz-xmaster** [comment]
Thanks for the help. After a few attempts, I still can't solve this [issue](https://travis-ci.com/github/open-mmlab/mmdetection/jobs/346492181), could you help with this problem?

**Comment by hyz-xmaster** [comment]
Excellent!

**Comment by hellock** [comment]
@yhcao6 will benchmark the new settings.

File: mmdet/models/dense_heads/fcos_head.py [hunk 2]
```diff
@@ -194,16 +208,24 @@ def loss(self,
...
-            loss_centerness = self.loss_centerness(pos_centerness,
-                                                   pos_centerness_targets)
+                avg_factor=pos_centerness_targets_avg_per_gpu)
+            loss_centerness = self.loss_centerness(
+                pos_centerness,
+                pos_centerness_targets,
+                avg_factor=num_pos_avg_per_gpu)
         else:
             loss_bbox = pos_bbox_preds.sum()
+            reduce_mean(pos_centerness.new_tensor([0.0]))
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/fcos_head.py:228]: We might need to send 1 for reduce in case all GPUS have no positive sample?

Comment (hyz-xmaster) [reply mmdet/models/dense_heads/fcos_head.py:228]: Well, I almost have no knowledge in DDP and just copied [that line](https://github.com/tianzhi0549/FCOS/blob/5fd728bfea150359fae28acb527e6e89d43c27e3/fcos_core/modeling/rpn/fcos/loss.py#L279) from the author of FCOS. So if you think it necessary to modify it, please feel free to do it.

**Comment by yhcao6** [comment]
Hi, @hyz-xmaster , I benchmark 6 models with the new setting, but the performance of single scale schedule all drops. I will re-benchmark these 4 models using the current setting.
| Model            | 4gpu | 8gpu |
|------------------|------|------|
| R-50 1x          | 36.6 | 36.3 |
| R-50 2x          | 36.9 | 36.5 |
| R-101 1x         | 39.2 | 38.9 |
| R-101 2x         | 39.1 | 38.5 |
| R-50 2x mstrain  | 38.7 | 38.9 |
| R-101 2x mstrain | 40.9 | 41.0 |

**Comment by hyz-xmaster** [comment]
> Hi, @hyz-xmaster , I benchmark 6 models with the new setting, but the performance of single scale schedule all drops. I will re-benchmark these 4 models using the current setting.
> 
> Model	4gpu	8gpu
> R-50 1x	36.6	36.3
> R-50 2x	36.9	36.5
> R-101 1x	39.2	38.9
> R-101 2x	39.1	38.5
> R-50 2x mstrain	38.7	38.9
> R-101 2x mstrain	40.9	41.0

That's a bit weird. I have only run the improved version of FCOS using 2 GPUs (8 imgs per gpu) and got 38.6 mAP, which is consistent with the author's result. Maybe one way of figuring this issue out is to run the author's implementation with different GPU settings to see if there is still a performance gap. If there is no such gap, that means something is wrong.

**Comment by hyz-xmaster** [comment]
> Hi, @hyz-xmaster , I benchmark 6 models with the new setting, but the performance of single scale schedule all drops. I will re-benchmark these 4 models using the current setting.
> 
> Model	4gpu	8gpu
> R-50 1x	36.6	36.3
> R-50 2x	36.9	36.5
> R-101 1x	39.2	38.9
> R-101 2x	39.1	38.5
> R-50 2x mstrain	38.7	38.9
> R-101 2x mstrain	40.9	41.0

Hi @yhcao6, 
   I just compared FCOS configs with ATSS and RepPoints's, and found that only FCOS is still using Caffe style pre-trained weights and its BN.requires_grad=False. I am not sure how this will impact the training with different GPUs. Do you have any insight ? Or replace the Caffe style weights with PyTorch style weights to see the results? Thanks.

**Comment by hyz-xmaster** [comment]
> Hi, @hyz-xmaster , I benchmark 6 models with the new setting, but the performance of single scale schedule all drops. I will re-benchmark these 4 models using the current setting.
> 
> Model	4gpu	8gpu
> R-50 1x	36.6	36.3
> R-50 2x	36.9	36.5
> R-101 1x	39.2	38.9
> R-101 2x	39.1	38.5
> R-50 2x mstrain	38.7	38.9
> R-101 2x mstrain	40.9	41.0

One interesting thing I just noticed is that in both of the implementations of RetinaNet in [maskrcnn benchmark](https://github.com/facebookresearch/maskrcnn-benchmark/blob/57eec25b75144d9fb1a6857f32553e1574177daf/maskrcnn_benchmark/modeling/rpn/retinanet/loss.py#L75) and [mmdetection](https://github.com/open-mmlab/mmdetection/blob/ca115f2d14b1a4235fdbe0e3f75fbe6e92149b35/mmdet/models/dense_heads/anchor_head.py#L353) this **num_pos synchronization trick** is not used at all. So is this synchronization correct  in theory and necessary in practice?

**Comment by yhcao6** [comment]
@hyz-xmaster,
I found the reason leads to the performance drops. In the latest commits, the default warm-up policy of FCOS had been changed slightly. Now I open a PR(#3119) to restore them. The performance should be able to align with 4GPUS now on 1x training schedules.

**Comment by hyz-xmaster** [comment]
> @hyz-xmaster,
> I found the reason leads to the performance drops. In the latest commits, the default warm-up policy of FCOS had been changed slightly. Now I open a PR(#3119) to restore them. The performance should be able to align with 4GPUS now on 1x training schedules.

Great to hear. The warmup_ratio and policy can have such a big impact on the performance? Is there any research on this?
